pkg/controller/usbdevice/usbdevice_claim_controller.go

```diff
@@ -0,0 +1,283 @@
+		}
+		return usbDeviceClaim, err
+	}
+
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	virt, err := h.virtClient.Get(KubeVirtNamespace, KubeVirtResource, metav1.GetOptions{})
+	if err != nil {
+		logrus.Errorf("failed to get kubevirt: %v", err)
+		return usbDeviceClaim, err
+	}
+
+	newVirt, err := h.updateKubeVirt(virt, usbDevice)
+	if err != nil {
+		logrus.Errorf("failed to update kubevirt: %v", err)
+		return usbDeviceClaim, err
+	}
+
+	// start device plugin if it's not started yet.
+	if _, ok := h.devicePlugin[usbDeviceClaim.Name]; ok {
+		return usbDeviceClaim, nil
+	}
+
+	pluginDevices := discoverAllowedUSBDevices(newVirt.Spec.Configuration.PermittedHostDevices.USB)
+
+	if pluginDevice := h.findDevicePlugin(pluginDevices, usbDevice); pluginDevice != nil {
+		usbDevicePlugin := h.devicePluginConvertor(usbDevice.Status.ResourceName, []*deviceplugins.PluginDevices{pluginDevice})
+		deviceHan := &deviceController{
+			device: usbDevicePlugin,
+		}
+		h.devicePlugin[usbDeviceClaim.Name] = deviceHan
+		h.startDevicePlugin(deviceHan)
+	}
+
+	usbDeviceCp := usbDevice.DeepCopy()
+	usbDeviceCp.Status.Enabled = true
+	if _, err = h.usbClient.UpdateStatus(usbDeviceCp); err != nil {
+		logrus.Errorf("failed to enable usb device %s status: %v", usbDeviceCp.Name, err)
+		return usbDeviceClaim, err
+	}
+
+	usbDeviceClaimCp := usbDeviceClaim.DeepCopy()
+	usbDeviceClaimCp.Status.PCIAddress = usbDevice.Status.PCIAddress
+	usbDeviceClaimCp.Status.NodeName = usbDevice.Status.NodeName
+
+	return h.usbClaimClient.UpdateStatus(usbDeviceClaimCp)
+}
+
+func (h *ClaimHandler) startDevicePlugin(deviceHan *deviceController) {
+	if deviceHan.started {
+		return
+	}
+
+	deviceHan.stop = make(chan struct{})
+
+	go func() {
+		if err := deviceHan.device.Start(deviceHan.stop); err != nil {
+			logrus.Errorf("failed to start device plugin: %v", err)
+		}
+		<-deviceHan.stop
+	}()
+
+	deviceHan.started = true
+}
+
+func (h *ClaimHandler) stopDevicePlugin(deviceHan *deviceController) error {
+	if !deviceHan.started {
+		return nil
+	}
+
+	close(deviceHan.stop)
+	deviceHan.started = false
+
+	if err := deviceHan.device.StopDevicePlugin(); err != nil {
+		logrus.Errorf("failed to start device plugin: %v", err)
+		return err
+	}
+
+	return nil
+}
+
+func (h *ClaimHandler) findDevicePlugin(pluginDevices map[string][]*deviceplugins.PluginDevices, usbDevice *v1beta1.USBDevice) *deviceplugins.PluginDevices {
+	var pluginDevice *deviceplugins.PluginDevices
+
+	for resourceName, devices := range pluginDevices {
+		for _, device := range devices {
+			device := device
+			for _, d := range device.Devices {
+				logrus.Debugf("resourceName: %s, device: %v", resourceName, d)
+				if usbDevice.Status.DevicePath == d.DevicePath {
+					pluginDevice = device
+					return pluginDevice
+				}
+			}
+		}
+	}
+
+	return pluginDevice
+}
+
+func (h *ClaimHandler) OnRemove(_ string, claim *v1beta1.USBDeviceClaim) (*v1beta1.USBDeviceClaim, error) {
+	if claim == nil {
+		return claim, nil
+	}
+
```

Comment:
we need a check to ignore claims without a deletion timestamp

Reply:
it's fixed, but didn't show out date .. weird